src/terminal/terminal.rs

```diff
@@ -545,51 +580,108 @@ where
             return Ok(());
         }
 
-        // Clear the viewport off the screen
-        self.clear()?;
-
-        // Move the viewport by height, but don't move it past the bottom of the terminal
-        let viewport_at_bottom = self.last_known_area.bottom() - self.viewport_area.height;
-        self.set_viewport_area(Rect {
-            y: self
-                .viewport_area
-                .y
-                .saturating_add(height)
-                .min(viewport_at_bottom),
-            ..self.viewport_area
-        });
-
-        // Draw contents into buffer
+        // Draw contents to insert into a buffer. The rest of this function is concerned with
+        // drawing this buffer onto the screen.
         let area = Rect {
-            x: self.viewport_area.left(),
+            x: 0,
             y: 0,
             width: self.viewport_area.width,
             height,
         };
         let mut buffer = Buffer::empty(area);
         draw_fn(&mut buffer);
-
-        // Split buffer into screen-sized chunks and draw
-        let max_chunk_size = (self.viewport_area.top() as usize) * (area.width as usize);
-        for buffer_content_chunk in buffer.content.chunks(max_chunk_size) {
-            let chunk_size = (buffer_content_chunk.len() / (area.width as usize)) as u16;
-
-            self.backend
-                .append_lines(self.viewport_area.height.saturating_sub(1) + chunk_size)?;
-
-            let iter = buffer_content_chunk.iter().enumerate().map(|(i, c)| {
-                let (x, y) = buffer.pos_of(i);
-                (
-                    x,
-                    self.viewport_area.top().saturating_sub(chunk_size) + y,
-                    c,
-                )
-            });
-            self.backend.draw(iter)?;
-            self.backend.flush()?;
-            self.set_cursor_position(self.viewport_area.as_position())?;
+        let mut buffer = buffer.content.as_slice();
+
+        // Use i32 variables so we don't have worry about overflowed u16s when adding, or about
+        // negative results when subtracting.
```

Comment:
Does it make sense to instead use saturating_add / sub methods for these, or are there intermediate points where something may go negative and then go back positive which would not work?

Reply:
It's the latter. They sometimes go over `u16::MAX`, and sometimes go negative. The latter is especially tricky because you want to be able to write `x + y - z` or `x - y + z` and not worry about intermediate results overflowing or becoming negative.